sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/instrumentation/ServiceBusMeter.java

```diff
@@ -0,0 +1,240 @@
+     * with {@link ServiceBusMeter#trackSettlementSequenceNumber()}.
+     */
+    public void reportSettlement(long start, long seqNo, DispositionStatus status, Throwable throwable, Context context) {
+        if (isEnabled) {
+            if (settleMessageDuration.isEnabled()) {
+                TelemetryAttributes attributes = throwable == null ? settleSuccessAttributes[status.ordinal()]
+                    : settleFailureAttributes[status.ordinal()];
+
+                settleMessageDuration.record(Instant.now().toEpochMilli() - start, attributes, context);
+            }
+
+            CompositeSubscription subs = lastSeqNoSubscription.get();
+            if (settledSequenceNumber.isEnabled() && subs != null) {
+                subs.set(seqNo, status, throwable == null);
+            }
+        }
+    }
+
+    /**
+     * Creates gauge subscription to report the latest sequence number value. Call it in client constructor and
+     * close once client closes.
+     */
+    AutoCloseable trackSettlementSequenceNumber() {
+        if (!isEnabled || !settledSequenceNumber.isEnabled()) {
+            return NOOP_CLOSEABLE;
+        }
+
+        CompositeSubscription subs = new CompositeSubscription();
+        lastSeqNoSubscription.set(subs);
+        return subs;
+    }
+
+    private class CompositeSubscription implements AutoCloseable {
+        private final AtomicLong[] lastSeqNoSuccess = new AtomicLong[DISPOSITION_STATUSES_COUNT];
+        private final AtomicLong[] lastSeqNoFailure = new AtomicLong[DISPOSITION_STATUSES_COUNT];
+        private final AutoCloseable[] subscriptionsSuccess = new AutoCloseable[DISPOSITION_STATUSES_COUNT];
+        private final AutoCloseable[] subscriptionsFailure = new AutoCloseable[DISPOSITION_STATUSES_COUNT];
+        private final Object lock = new Object();
+
+        CompositeSubscription() {
+            for (int i = 0; i < DISPOSITION_STATUSES_COUNT; i++) {
+                lastSeqNoSuccess[i] = new AtomicLong();
+                lastSeqNoFailure[i] = new AtomicLong();
+            }
+        }
+
+        public void set(long value, DispositionStatus status, boolean success) {
+            getOrCreate(status, success).set(value);
+        }
+
+        @Override
+        public void close() {
+            for (int i = 0; i < subscriptionsSuccess.length; i++) {
+                if (subscriptionsSuccess[i] != null) {
+                    try {
+                        subscriptionsSuccess[i].close();
+                        subscriptionsFailure[i].close();
+                    } catch (Exception ex) {
+                        LOGGER.info("Unable to close settlement sequence number subscription.", ex);
+                    }
+                }
+            }
+        }
+
+        private AtomicLong getOrCreate(DispositionStatus status, boolean success) {
+            final int i = status.ordinal();
+            if (success) {
+                if (subscriptionsSuccess[i] == null) {
```

Comment:
The DCL solves the visibility issue for the variable in question. For example, if we want to initialize a variable `f`, then declaring `f` as volatile + DCL ensures the initialization is immediately visible to all threads accessing that memory.

But if the variable is an array, as in `f[]`, then (my understanding is) volatile + DCL on the array won't make the reads and writes to elements of the array volatile.

The below blog is a write-up on this topic (volatile array)

[Java Concurrency (&c): Volatile Arrays in Java (jeremymanson.blogspot.com)](http://jeremymanson.blogspot.com/2009/06/volatile-arrays-in-java.html)

It seems complicated, but I'm unaware of a variant of DCL that fully ensures visibility of array elements write/read. Since I lack a point of reference, I thought at least share what I know. But If you have already come across more details guaranteeing this correctness, please let me know. I love to learn.

That said, I wonder, in the long run (typical case of SB/EH), how much we save with this lazy initialization of array over eagerly init it in Ctr when tracing enabled? 

It looks like there are 4-5 known disposition statuses (enum, so it likely not change as it's a breaking change), then multiply that by 2 (for success vs. failure), total ~10 objects, which is anyway expected to be allocated in the long run.

Please share your thoughts.

Reply:
great point! you're right - we're not gaining much with lazy initialization and that array initialization synchronization is sketchy. I change it to create all subscriptions in ctor. 
The subtle difference in behavior is that if there were no messages dead-lettered, users would see 0, before they would not see any data. I think it's ok.

Reply:
thanks for checking.